File: python/4-array/8-mengelompokan-hewan.py
def groupAnimals(animals):
  sorted_animal = sorted(animals)
  result = []
  mark = sorted_animal[0][0]

  grouped = []
  for sa in sorted_animal:
    if sa[0] == mark:
      grouped.append(sa)
    else:
      result.append(grouped)
      grouped = []
      grouped.append(sa)
      mark = sa[0]
  result.append(grouped)
  return result

# Driver Code
print(groupAnimals(['cacing', 'ayam', 'kuda', 'anoa', 'kancil']))
# [ ['ayam', 'anoa'], ['cacing'], ['kuda', 'kancil'] ]
print(groupAnimals(['cacing', 'ayam', 'kuda', 'anoa', 'kancil', 'unta', 'cicak' ]))
# [['anoa', 'ayam'], ['cacing', 'cicak'], ['kancil', 'kuda'], ['unta]]

File: go/4_basic_array/8_mengelompokan_hewan.go
package main

import (
	"fmt"
	"sort"
)

func GroupingAnimal(animals []string) [][]string {
	sort.Strings(animals)
	result := [][]string{}

	mark := string([]rune(animals[0])[0])

	grouped := []string{}
	for _, animal := range animals {
		curMark := string(animal[0])
		if curMark == mark {
			grouped = append(grouped, animal)
		} else {
			result = append(result, grouped)
			grouped = []string{}
			grouped = append(grouped, animal)
			mark = string(animal[0])
		}
	}
	result = append(result, grouped)
	return result
}

func main() {
	fmt.Println(GroupingAnimal([]string{"cacing", "ayam", "kuda", "anoa", "kancil"}))
	// [['ayam', 'anoa'], ['cacing'], ['kuda', 'kancil'] ]
	fmt.Println(GroupingAnimal([]string{"cacing", "ayam", "kuda", "anoa", "kancil", "unta", "cicak"}))
	// [['anoa', 'ayam'], ['cacing', 'cicak'], ['kancil', 'kuda'], ['unta]]
}
